### PR TITLE
Add token lifetime constraints to JwtProvider

### LINKED ISSUE #31455: JWT-SVID support for jwt_authn filter
*JWT-SVID restrictions*: *Add configuration options for `jwt_authn` to implement the JWT-SVID spec*

*Description*:
Based on the JWT-SVID standard, the jwt_authn filter will accept JWT-SVIDs today, but won't apply stricter requirements of the spec. Based on the standard, I believe the gaps are the following:

1. [3.1 Subject](https://github.com/spiffe/spiffe/blob/main/standards/JWT-SVID.md#31-subject): jwt_authn doesn't have restrictions on the `sub` claims for a JWT provider. To meet the standard, we should only allow SPIFFE IDs for the subject
2. [3.3 Expiration Time](https://github.com/spiffe/spiffe/blob/main/standards/JWT-SVID.md#33-expiration-time): jwt_authn needs a way to require an expiration time
3. [6.1 Publishing SPIFFE Bundle Elements](https://github.com/spiffe/spiffe/blob/main/standards/JWT-SVID.md#61-publishing-spiffe-bundle-elements): `JwtProvider` does not provide a way to require a specific `use` paramater for a JWK entry. Per the spec, only `jwt-svid` is allowed
4. [6.1 Publishing SPIFFE Bundle Elements](https://github.com/spiffe/spiffe/blob/main/standards/JWT-SVID.md#61-publishing-spiffe-bundle-elements): There's no way to specify that a given `JwtProvider` can only sign JWTs for a specific trust domain

My general proposal is to add some additional configuration to `jwt_authn` for general restrictions and maybe provide a convenience JWT-SVID configuration. The proposed changes are:

**`sub` restrictions**:
For a given `JwtProvider` add a `subjects` field similar to `audiences`. The values would be a list of acceptable subjects with wildcards. Subjects that don't match a requirement would be rejected. This also solves problem 4 above by restricting the trust domain of the subject.

Example:
```yaml
issuer: spiffe://trustdomain.example.com
subjects:
- spiffe://trustdomain.example.com/*
```

**`exp` restrictions**:
For a given `JwtProvider` add a `max_expiration` field that would be duration representing the maximum expiration of a JWT. We would need a special value for any expiration, as long as there is one.

Example:
```yaml
issuer: spiffe://trustdomain.example.com
max_expiration: 4h
```

**`use` restrictions**:
For a given `JwtProvider` add a `uses` field similar to `audiences`. This would be a filter on JWK entries in the specified JWKS.

Example
```yaml
issuer: spiffe://trustdomain.example.com
uses:
- jwt-svid
```

**Convenience config**:
To help users configure the filter correctly, we could provide a configuration similar to `extensions.transport_sockets.tls.v3.SPIFFECertValidatorConfig`. The config could be part of the `JwtProvider` or if needed we could create a new `SpiffeJwtProvider`.

Example JwtProvider extension:
```yaml
spiffe_trust_domain: trustdomain.example.com #required
max_expiration: 4h #optional, defaults to present
remote_jwks: ...
```

Would be equivalent to:
```yaml
issuer: spiffe://trustdomain.example.com
max_expiration: 4h
uses:
- jwt-svid
subjects:
- spiffe://trustdomain.example.com/*
```

*Relevant Links*:
* https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/tls_spiffe_validator_config.proto
* https://github.com/spiffe/spiffe/blob/main/standards/JWT-SVID.md

### PR DESCRIPTION
**Pull request by mattjo**
Commit Message: Add max token lifetime constraint for JwtProviders.
Additional Description: Adds new max_lifetime config field to restrict token lifetime accepted from a JwtProvider.
Risk Level: Low
Testing: Unit testing
Docs Changes: Added subjects description inline in proto.
Release Notes: Attached
Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md): Feature is opt in, without specifying the config, there's no behavior change.
Fixes #31455

### CODE REVIEW

**Comment by mattjo** [comment]
Similar to #32374, but adds maximum token lifetime constraints.

**Comment by mattjo** [comment]
@lizan and @abeyad this should be similar to https://github.com/envoyproxy/envoy/pull/32374, but adds maximum token lifetime constraints. Thanks for your review!

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto [hunk 1]
```diff
@@ -119,6 +119,18 @@ message JwtProvider {
...
+  // are not valid for more than one hour. Credentials valid for longer will fail validation. Setting this to inf will
+  // require that an expiration is set in the JWT, but not check the lifetime.
+  //
+  // Example:
+  //
+  // .. code-block:: yaml
+  //
+  //     max_lifetime: 1h
+  //
+  string max_lifetime = 20;
```
[the remark this answers is not in the document]
Comment (abeyad) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:132]: +1, this should be google.protobuf.Duration

[the remark this answers is not in the document]
Comment (mattjo) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto]: Pushed a new commit with the change described, lmk how it looks

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto [hunk 2]
```diff
@@ -119,6 +119,31 @@ message JwtProvider {
   //
   type.matcher.v3.StringMatcher subjects = 19;
 
+  oneof lifetime_constraint {
+    // Restrict the maximum remaining lifetime of a credential from the JwtProvider. For instance, to make sure that tokens
+    // are not valid for more than one hour. Credentials valid for longer will fail validation. Setting this to inf will
+    // require that an expiration is set in the JWT, but not check the lifetime.
```
Comment (abeyad) [inline api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:125]: > Setting this to inf will require that an expiration is set in the JWT, but not check the lifetime.

I'm not clear on what this means.  It sounds like the JWT token can have an expiration, and you can also set a lifetime here, and if you set a lifetime here w/ max_lifetime, then it'll override the JWT token expiration, but if you set max_lifetime to `inf`, then the JWT token expiration will be used instead?

Comment (mattjo) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:125]: Let me rephrase that. What we're trying to capture is that if you set a max_lifetime of 24 hours, then tokens that are valid for longer than that will be rejected. This is to stop the case where an issuer issues tokens that are valid for a long period and don't have a way to revoke them if they're stolen.

But, the JWT-SVID [spec](https://github.com/spiffe/spiffe/blob/main/standards/JWT-SVID.md#33-expiration-time) only requires that there is an expiration, but doesn't require a specific value. So the idea is to let people configure their proxies to meet the minimum of the spec or be more strict.

Comment (mattjo) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto]: Redid the docs

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto [hunk 3]
```diff
@@ -119,6 +119,35 @@ message JwtProvider {
...
+    // `expiration restrictions <https://github.com/spiffe/spiffe/blob/main/standards/JWT-SVID.md#33-expiration-time>`_.
+    // Unlike ``max_lifetime``, this only requires that expiration is present, where ``max_lifetime`` also checks the value.
+    //
+    // Example:
+    //
+    // .. code-block:: yaml
+    //
+    //     require_expiration: true
+    //
+    bool require_expiration = 21;
```
Comment (abeyad) [inline api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:148]: Thanks, this makes it much clearer!  In general, the new guidance is to avoid `oneof`, see https://github.com/envoyproxy/envoy/blob/3c84a245c056ae5bd74309d766091f33d9190071/api/STYLE.md?plain=1#L63.

If you want to require one or the other (either max_lifetime must be set, or require_expiration must be set), why not enforce that in the code?  I.e. just have `max_lifetime` field... if it is not set, the code checks that the JWT has an expiration on it; if not, raise an error condition.

Comment (mattjo) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:148]: Thanks for the link, very helpful. There's a 3rd case where neither is set because the JWT spec does not require expiration for a token. And, envoy does not currently require one, so setting nothing and implying require_expiration would change behavior for existing deployments. But, based on the guide I think this is very similar to the example where max_lifetime is a superset of require_expiration. We could also have some special value in the duration, like setting duration to 0 for require any expiration, but I think that's not as clear. Let me try one more.

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto [hunk 4]
```diff
@@ -119,6 +119,34 @@ message JwtProvider {
...
+  // over ``require_expiration``.
+  //
+  // Example:
+  //
+  // .. code-block:: yaml
+  //
+  //     max_lifetime:
+  //       seconds: 86400
+  //
+  google.protobuf.Duration max_lifetime = 21;
```
Comment (abeyad) [inline api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:148]: looks great, thanks for making the change!

**Review by abeyad — DISMISSED** [review]
/lgtm api

File: changelogs/current.yaml [hunk 5]
```diff
@@ -149,6 +149,11 @@ bug_fixes:
 - area: jwt_authn
   change: |
     Fixed JWT extractor, which concatenated headers with a comma, resultig in invalid tokens.
+- area: jwt_authn
+  change: |
+    Added
+    :ref:`max_lifetime <envoy_v3_api_field_extensions.filters.http.jwt_authn.v3.JwtProvider.max_lifetime>` to limit
+    the maximum remaining lifetime of a token from a ``JwtProvider``.
```
Comment (abeyad) [inline changelogs/current.yaml:156]: nit: do you also want to add a note about the require_expiration field here?

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto (hunk 4 above)
Comment (mattjo) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:148]: Thanks for your comments, I agree, looks better now!

**Review by abeyad — APPROVED** [review]
/lgtm api

**Review by lizan — APPROVED** [review]
